### PR TITLE
feat: add mod update checker

### LINKED ISSUE #84: Mod update checker
- Uses Curseforge API to check current mods' new releases for launched game version 
- Checks once per game session 
- If there are updates, show the Realms 🔷 icon on mod list button 
- Checks for stable releases by default, option to use beta/alpha too 
- "Update" badge on mods that have updates 
- A button "Update mod" on the mod detail view that directly links to latest JAR on Curseforge

### PR DESCRIPTION
**Pull request by JustPyrrha**
Closes #84 

example of mod config (fabric.mod.json)
```json
{
  "custom": {
    "modmenu": {
      "updates": {
        "provider": "modrinth",
        "projectId": "mOgUt4GM"
      }
    }
  }
}
```

Tasks:
- [x] Update Providers
  - [x] Modrinth.
  ```json
  {
    "provider": "modrinth",
    "projectId": "mOgUt4GM"
  }
  ```
  - [x]  ~~CurseForge.~~ (see [this comment](https://github.com/TerraformersMC/ModMenu/pull/235#issuecomment-806383667).)
  - [x]  Maven.
  ```json
  {
    "provider": "maven",
    "repository": "https://maven.example.com/",
    "group": "com.example",
    "artifact": "ExampleMod",
    "versionRegEx": "([a-zA-Z0-9]\\S+)(\\+mc1\\.16\\.5)"
  }
  ```
  - [x]  Github.
  ```json
  {
    "provider": "github",
    "repository": "owner/repo",
    "allowPrerelease": false,
    "versionRegEx": "([a-zA-Z0-9]\\S+)(\\+mc1\\.16\\.5)"
  }
  ```
- [x] Display outdated.
  - [x] Title Corner.
  
  - [x] Mods Button.
  
  - [x] Mods Button (slim).
  
- [x] Display update info. 
  
- [x] ~~Outdated badge.~~ (colors not final)
  
  Replaced with icon in mod list.
  
- [x] Log outdated mods.
  
- [x] Properly dispose HTTP clients.
- [x] Disable update checks config option.
  
- [x] Dummy data for fabric loader

New keys to translate
```json
{
  "modmenu.loaded.outdated": "(%s Loaded, %s Outdated)",
  "modmenu.updateAvailable": "An update is available. (v%s -> v%s)",
  "modmenu.downloadFrom": "Download v%s from:",
  "modmenu.changelog": "Changes in v%s:",
  "modmenu.fabricmcNet": "FabricMC.net",
  "modmenu.mods.1.outdated": " (%s Mod, %s Outdated)",
  "modmenu.mods.69.nice.outdated": " (%s Mods...nice, %s Outdated)",
  "modmenu.mods.a.outdated": " (%s Mods, %s Outdated)",

  "option.modmenu.disable_update_checks": "Update Checks: %s",
  "option.modmenu.disable_update_checks.false": "Enabled",
  "option.modmenu.disable_update_checks.true": "Disabled"
}
```

### CODE REVIEW

**Comment by Prospector** [comment]
Thanks for starting this! One thing I wonder about is how you plan on doing it on CF, since it doesn't actually have a mod version field. Plus using their "API" is not something they support.

**Comment by JustPyrrha** [comment]
oh, I didn't realize that curse doesn't support their API. We might have to do some page scraping :)

**Comment by Prospector** [comment]
Well, at least it breaks very often and despite having some public documentation on it, they refer to it as their "internal APIs"

I would lean more towards simply not supporting this feature for CurseForge tbh

**Comment by JustPyrrha** [comment]
> I would lean more towards simply not supporting this feature for CurseForge tbh

Yeah, that's probably for the best.

Edit: PR description updated accordingly.

**Comment by Wyn-Price** [comment]
The only way I imagine it possible with the curse API, it to query the project files and check the newest file for the mc version being played, and compare the file name to the jar file name.

As for the stability of the API, it's been unchanging since I first heard about it a year an a half ago, with the only hiccup being when the servers changed in December. 

This isn't a particularly fun solution but as curseforge is the most popular modding platform (for now), it would make sense to add support, even if it's just checking the jar name. I've done a fair bit of work with the curse API before so if you wanted help implementing it I'm available.

**Comment by JustPyrrha** [comment]
this should be ready for review now

**Comment by JustPyrrha** [comment]
I gotta say, this my favorite comment from looking through the mod's code.
image

File: src/main/resources/assets/modmenu/lang/en_us.json [hunk 1]
```diff
@@ -3,6 +3,8 @@
   "modmenu.descriptionTranslation.modmenu": "Adds a mod menu to view the list of mods you have installed.",
   "modmenu.loaded": "(%s Loaded)",
   "modmenu.loaded.short": "(%s)",
+  "modmenu.loaded.outdated": "(%s Loaded, %s Outdated)",
+  "modmenu.loaded.short.outdated": "(%s,%s)",
```
Comment (Prospector) [inline src/main/resources/assets/modmenu/lang/en_us.json]: I feel like this is pretty ambiguous and confusing, just having 2 numbers there like that. Maybe for the short mod count we just omit the outdated number?

Comment (JustPyrrha) [reply src/main/resources/assets/modmenu/lang/en_us.json]: fixed in 858c438

File: src/main/java/com/terraformersmc/modmenu/updates/providers/GithubUpdateProvider.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	private static final Gson gson = new GsonBuilder().create();
+
+	public GithubUpdateProvider(String gameVersion) {
+		super(gameVersion);
+	}
+
+	@Override
+	public void check(String modId, String version, FabricMod.ModUpdateData data, Consumer<AvailableUpdate> callback) {
+		beginUpdateCheck();
+		Thread thread = new Thread(String.format("Update Checker %s@Github", modId)) {
```
Comment (Wyn-Price) [inline src/main/java/com/terraformersmc/modmenu/updates/providers/GithubUpdateProvider.java]: You shouldn't really create a new thread every time. Try using something like `Util.getMainWorkerExecutor` to execute the request.

**Comment by Prospector** [comment]
> The only way I imagine it possible with the curse API, it to query the project files and check the newest file for the mc version being played, and compare the file name to the jar file name.
> 
> As for the stability of the API, it's been unchanging since I first heard about it a year an a half ago, with the only hiccup being when the servers changed in December.
> 
> This isn't a particularly fun solution but as curseforge is the most popular modding platform (for now), it would make sense to add support, even if it's just checking the jar name. I've done a fair bit of work with the curse API before so if you wanted help implementing it I'm available.

Hmm, maybe it would be good to have a primitive update checker for CF. I *would* sorta like to incentivize people to host their mods on multiple platforms such as Modrinth which aren't horribly stagnant and lacking in public and supported APIs, but it is true that many mods aren't on other platforms at the moment. Though, a pretty significant number of fabric mods *are* on Modrinth. Another thing to consider is that CurseForge already has an update checker built into their app for CF mods. In addition, even if most people get their mods through CurseForge, if the mod is also available on Modrinth, they can still use that as the update service.

I wouldn't be opposed to having it added, but I'd probably heavily encourage people to use another provider in the documentation, and if it were to break I might just drop it.

File: src/main/java/com/terraformersmc/modmenu/updates/providers/GithubUpdateProvider.java (hunk 2 above)
Comment (JustPyrrha) [reply src/main/java/com/terraformersmc/modmenu/updates/providers/GithubUpdateProvider.java]: fixed in 08d0bf2

**Comment by JustPyrrha** [comment]
Not sure if this is all endpoints but the `files` endpoint (`https://minecraft.curseforge.com/api/projects/308702/files/`) requires an API token. Not sure if there's a way to poll the API without a token.
IMO this rules out supporting curse, I don't want to bundle an API token with the mod, mainly for the security concerns with publishing an API token.

**Comment by Wyn-Price** [comment]
Use https://twitchappapi.docs.apiary.io/#

**Comment by JustPyrrha** [comment]
> Use https://twitchappapi.docs.apiary.io/#

Seems to have been removed, might be dependent on [CF-I-10](https://curseforge-ideas.overwolf.com/ideas/CF-I-10) now.

**Review by hxr404 — APPROVED** [review]
thanks its super useful!
Will this be merged?

**Comment by JustPyrrha** [comment]
> Will this be merged?

I think so, it's be a requested feature for quite some time now. If we're not going to wait for the Curse api it should be fairly soon.

**Comment by hxr404** [comment]
Is this fork fully compatible with mods that use [this](https://gitea.thebrokenrail.com/TheBrokenRail/ModUpdater/src/branch/master/MOD_DEVELOPER.md) format?
Many Mods have opted-in for [TheBrokenRail/ModUpdater](https://gitea.thebrokenrail.com/TheBrokenRail/ModUpdater) or work in combination with [FabricZero](https://github.com/Fox2Code/FabricZero)
So are the updating features of those two mods also implemented?
Than we wouldn't really need CurseForge anymore (cuz most mods are also on GitHub)

**Comment by JustPyrrha** [comment]
> Is this fork fully compatible with mods that use this format?

The format looks similar but isnt the same, for example this doesnt have the json strategy provided by ModUpdater.

> So are the updating features of those two mods also implemented?

No, this only notifies the user that an update is available and lets them know where they can download them from.

> Than we wouldn't really need CurseForge anymore (cuz most mods are also on GitHub)

It's still a platform that mod developers use so it should be considered.

**Comment by Wyn-Price** [comment]
If needed I can add the current curse api to it. It still exists the docs site just seems to be broken. I know the api well enough to help.

**Comment by Wyn-Price** [comment]
Also: https://gaz492.github.io/CurseforgeAPI/

**Comment by JustPyrrha** [comment]
> If needed I can add the current curse api to it. It still exists the docs site just seems to be broken. I know the api well enough to help.

Go for it. I don't think I have enough experience with curse in general to do it anyway.

**Comment by spnda** [comment]
@JoeZwet What's the current state of this? https://github.com/JoeZwet/ModMenu/pull/2 is still open and for 25 days there has been no review/feedback towards that pr.
I also personally think undocumented API should only be used with a "experimental" toggle or similar, but a Curse integration is definitely a positive addition.

**Comment by NeusFear** [comment]
I would like this to be based on our 1.17 branch. We also need some more updated reviews for the latest commits.

File: src/main/resources/fabric.mod.json [hunk 3]
```diff
@@ -96,9 +96,9 @@
         "modmenu.discord": "https://discord.gg/jEGF5fb"
       },
       "updates": {
-        "provider": "modrinth",
-        "projectId": "mOgUt4GM",
-        "channel": "release"
+        "provider": "curseforge",
```
Comment (luximus-hunter) [inline src/main/resources/fabric.mod.json:102]: Does this mean the update provider for ModMenu will be CF? If so, why?

**Comment by luximus-hunter** [comment]
Would it be possible to make the update arrow green, as that usually signifies a download

image
Untitled

**Comment by spnda** [comment]
I just rebased this fork onto the [1.17 branch](https://github.com/spnda/ModMenu/tree/update-checker-1.17), if anyone wants to merge that instead to get this merged? It's just been going a little slow recently and I'd be really stoked if this got merged and adopted my more mod authors sooner. I'd really be happy to write some extras based on reviews.

**Comment by luximus-hunter** [comment]
> I just rebased this fork onto the [1.17 branch](https://github.com/spnda/ModMenu/tree/update-checker-1.17), if anyone wants to merge that instead to get this merged? It's just been going a little slow recently and I'd be really stoked if this got merged and adopted my more mod authors sooner. I'd really be happy to write some extras based on reviews.

If it all works, perhaps make a new pull request on this repo merging the changes for it

**Comment by JustPyrrha** [comment]
My bad, been caught up with work. I'll sort this out in a bit.

**Comment by luximus-hunter** [comment]
> My bad, been caught up with work. I'll sort this out in a bit.

Take your time pal, personal health goes first

File: src/main/resources/fabric.mod.json (hunk 3 above)
Comment (JustPyrrha) [reply src/main/resources/fabric.mod.json:102]: I think this was left in from when the cf provider was added. Might be worth switching it back to modrinth.

**Comment by JustPyrrha** [comment]
> Would it be possible to make the update arrow green, as that usually signifies a download
> 
> image
> Untitled

The pr checks for updates, not downloads them. The icon is to show the mod is outdated.

**Comment by haykam821** [comment]
I would separate the translations for mod counts into the two components (mods and outdated).

**Comment by JustPyrrha** [comment]
> I would separate the translations for mod counts into the two components (mods and outdated).

They already are, I've only listed the ones added by the pr, there are already keys for mod counts [here](https://github.com/TerraformersMC/ModMenu/blob/1.17/src/main/resources/assets/modmenu/lang/en_us.json#L4-L7).

**Comment by haykam821** [comment]
The `%s Mods Loaded` and `%s Outdated` should be separate. That way, the mods loaded text is not dependent on the outdated text.

**Comment by JustPyrrha** [comment]
ah, I see what you mean, I'll do that after the rebase

**Comment by JustPyrrha** [comment]
:pain2: i broke it

**Comment by JustPyrrha** [comment]
and its fixed, fun

**Comment by K0-RR** [comment]
So, what about this?

**Comment by JustPyrrha** [comment]
ill rebase this onto 1.18, guess i should remove the cf checker too since the new api requires a key to do anything

**Comment by JustPyrrha** [comment]
just use #315

**Pull request closed without merging**